syntax = "proto3";

// =====================================================================
// Authenticate Service
// =====================================================================

message AuthenticateUserRequest {
    string email = 1;
    string password = 2;
}

message AuthenticateUserResponse {
    bool success = 1;
}

service AuthService {
    rpc AuthenticateUser (AuthenticateUserRequest) returns (AuthenticateUserResponse);
}

// =====================================================================
// GET ALL Students Service
// =====================================================================

message StudentDataRequest {}

message StudentUpdateRequest {
    int32 id = 1;
    string name = 2;
    int32 gender_id = 3;
    string birthday = 4;
    string email = 5;
    string phone = 6;
    int32 state_id = 7;
    string city = 8;
    string neighborhood = 9;
    string address = 10;
    string number = 11;
    string cep = 12;
    int32 payment_id = 13;
    int32 nivel_id = 14;
    int32 goal_id = 15;
}

message StudentUpdatePaymentIdRequest {
    int32 student_id = 1;
    int32 financial_id = 2;
}

message StudentCreateRequest {
    string name = 1;
    int32 gender_id = 2;
    string birthday =3;
    string email = 4;
    string phone = 5;
    int32 state_id = 6;
    string city = 7;
    string neighborhood = 8;
    string address = 9;
    string number = 10;
    string cep = 11;
    int32 payment_id = 12;
    int32 nivel_id = 13;
    int32 goal_id = 14;
}

message StudentSearchRequest {
    string student_name = 1;
}

message Student {
    int32 id = 1;
    string name = 2;
    string gender_id = 3;
    string birthday = 4;
    string email = 5;
    string phone = 6;
    string state_id = 7;
    string city = 8;
    string neighborhood = 9;
    string address = 10;
    string number = 11;
    string cep = 12;
    string payment_id = 13;
    string nivel_id = 14;
    string goal_id = 15;
}

message StudentDeleteRequest {
    int32 student_id = 1;
}

message StudentDataResponse {
    repeated Student student = 1;
}

message StudentUpdateResponse {
    bool success = 1;
}

message StudentCreateResponse {
    int32 id = 2;
}

service StudentService {
    rpc GetStudentData (StudentDataRequest) returns (StudentDataResponse);
    rpc UpdateStudentData (StudentUpdateRequest) returns (StudentUpdateResponse);
    rpc CreateStudentData (StudentCreateRequest) returns (StudentCreateResponse);
    rpc UpdateStudentPaymentId (StudentUpdatePaymentIdRequest) returns (StudentUpdateResponse);
    rpc DeleteStudentById (StudentDeleteRequest) returns (StudentUpdateResponse);
    rpc GetStudentByName (StudentSearchRequest) returns (StudentDataResponse);
}

// =====================================================================
// GET ALL States Service
// =====================================================================

message StatesDataRequest {}

message StateDataRequestById {
    int32 state_id = 1;
}

message StateResponse {
    int32 id = 1;
    string name = 2;
    string abbreviation = 3;
}

message StatesListDataResponse {
    repeated StateResponse state = 1;
}

service StateService {
    rpc GetStateData (StatesDataRequest) returns (StatesListDataResponse);
    rpc GetStateDataById (StateDataRequestById) returns (StateResponse);
}

// =====================================================================
// GET ALL Genders Service
// =====================================================================

message GendersDataRequest{}

message GendersDataRequestById{
    int32 gender_id = 1;
}

message GenderResponse {
    int32 id = 1;
    string name = 2;
}

message GenderListDataResponse {
    repeated GenderResponse gender = 1;
}

service GenderService {
    rpc GetGenderData (GendersDataRequest) returns (GenderListDataResponse);
    rpc GetGenderDataById (GendersDataRequestById) returns (GenderResponse);
}

// =====================================================================
// GET ALL Nivel Service
// =====================================================================

message NivelDataRequest{}

message NivelDataRequestById{
    int32 nivel_id = 1;
}

message NivelResponse {
    int32 id = 1;
    string name = 2;
}

message NivelListDataResponse {
    repeated NivelResponse nivel = 1;
}

service NivelService {
    rpc GetNivelData (NivelDataRequest) returns (NivelListDataResponse);
    rpc GetNivelDataById (NivelDataRequestById) returns (NivelResponse);
}

// =====================================================================
// GET ALL Goals Service
// =====================================================================

message GoalDataRequest{}

message GoalDataRequestById{
    int32 goal_id = 1;
}

message GoalResponse {
    int32 id = 1;
    string name = 2;
}

message GoalListDataResponse {
    repeated GoalResponse goal = 1;
}

service GoalService {
    rpc GetGoalsData (GoalDataRequest) returns (GoalListDataResponse);
    rpc GetGoalDataById (GoalDataRequestById) returns (GoalResponse);
}